sw/host/opentitanlib/src/test_utils/bitbanging.rs

```diff
@@ -0,0 +1,269 @@
+                    0x00 << SDA | 0x01 << SCL,
+                    0x01 << SDA | 0x00 << SCL,
+                    0x01 << SDA | 0x01 << SCL,
+                ],
+                Transfer::Addr(byte, read) => self.byte::<SDA, SCL>((byte << 1) | read as u8, read),
+                Transfer::Write(bytes) => bytes
+                    .iter()
+                    .flat_map(|byte| self.byte::<SDA, SCL>(*byte, false))
+                    .collect(),
+                Transfer::WriteNibble(byte) => self
+                    .byte::<SDA, SCL>(byte, false)
+                    .into_iter()
+                    .take(4 * 2)
+                    .collect(),
+                Transfer::Read(len) => (0..len)
+                    .flat_map(|index| {
+                        self.byte::<SDA, SCL>(0xff, index != (len - 1) || is_next_stop)
+                    })
+                    .collect(),
+                Transfer::Byte(byte, nak) => self.byte::<SDA, SCL>(byte, !nak),
+                Transfer::Bytes(bytes) => bytes
+                    .iter()
+                    .flat_map(|byte| self.byte::<SDA, SCL>(*byte, false))
+                    .collect(),
+            }
+        }
+
+        fn byte<const SDA: u8, const SCL: u8>(&self, byte: u8, ack: bool) -> Vec<u8> {
+            let mut buffer = (0..8u8)
+                .rev()
+                .flat_map(|bit| {
+                    [
+                        (((byte >> bit) & 0x01) << SDA) | 0x01 << SCL,
+                        (((byte >> bit) & 0x01) << SDA) | 0x00 << SCL,
+                    ]
+                })
+                .collect::<Vec<u8>>();
+            buffer.append(&mut self.ack::<SDA, SCL>(!ack));
+            buffer
+        }
+
+        fn ack<const SDA: u8, const SCL: u8>(&self, nack: bool) -> Vec<u8> {
+            vec![
+                (nack as u8) << SDA | 0x01 << SCL,
+                (nack as u8) << SDA | 0x00 << SCL,
+            ]
+        }
+    }
+    pub struct Encoder<const SDA: u8, const SCL: u8> {}
+    impl<const SDA: u8, const SCL: u8> Encoder<SDA, SCL> {
+        pub fn run(&self, transfer: &[Transfer]) -> Vec<u8> {
+            let mut result: Vec<u8> = transfer
+                .windows(2)
+                .flat_map(|window| {
+                    window[0].parse::<SDA, SCL>(window.get(1) != Some(&Transfer::Stop))
+                })
+                .collect();
+
+            //We missed the last element for using the windows function to peek, so we parse the last element here.
+            result.append(&mut transfer.iter().last().unwrap().parse::<SDA, SCL>(false));
+            result
+        }
+    }
+
+    enum DecodingStage {
+        Start,
+        Bytes,
+    }
+
+    pub struct Decoder<const SDA: u8, const SCL: u8> {
+        pub buffer: [u8; 256],
+    }
+
+    impl<const SDA: u8, const SCL: u8> Decoder<SDA, SCL> {
+        const LOW: u8 = 0;
+        const HIGH: u8 = 1;
+
+        fn sda(sample: u8) -> u8 {
+            (sample >> SDA) & 0x01
```

Comment:
This is a bit confusing. Perhaps create a sample struct that either:
* contains two field, sda and scl, or
* contains just a u8 and expose sda/scl methods.

Reply:
You should be able to completely eradicate the const generic approach by first encoding to `Vec<Sample>`, and then `samples.into_iter().map(Sample::encode::<0, 1>).collect()`.

Reply:
Done, looks much better now. Thanks!